### PR TITLE
[CARBONDATA-547] Added CarbonSession and enabled parser to use carbon custom commands

### PR DESCRIPTION
**Pull request by ravipesala**
This PR added CarbonSession and uses additional parser to parse the commands which are not supported by original Spark parser.

How to create CarbonSession?

```
 // First import the implicits like below
  import org.apache.spark.sql.CarbonSession._
    val spark = SparkSession
      .builder()
      .master("local")
      .enableHiveSupport()
      .getOrCreateCarbonSession()
     // Use getOrCreateCarbonSession to create the CarbonSession
```
It enables to support the `CREATE`, `LOAD`, `DROP`, `ALTER COMPACT` etc custom commands.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/245/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/280/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/281/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/318/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonExample.scala [hunk 1]
```diff
@@ -51,7 +52,7 @@ object CarbonExample {
         .config("spark.sql.warehouse.dir", warehouse)
         .config("javax.jdo.option.ConnectionURL",
           s"jdbc:derby:;databaseName=$metastoredb;create=true")
-        .getOrCreate()
+        .getOrCreateCarbonSession()
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonExample.scala]: Do not modify this and change file name to SparkSessionExample. And change CarbonExampleNew to CarbonSessionExample.

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonExampleNew.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      .config("javax.jdo.option.ConnectionURL",
+        s"jdbc:derby:;databaseName=$metastoredb;create=true")
+      .getOrCreateCarbonSession()
+    spark.sparkContext.setLogLevel("WARN")
+    CarbonProperties.getInstance()
+      .addProperty("carbon.kettle.home", s"$rootPath/processing/carbonplugins")
+      .addProperty("carbon.storelocation", storeLocation)
+
+    // Specify timestamp format based on raw data
+    CarbonProperties.getInstance()
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonExampleNew.scala]: set it with previous CarbonProperties in line 46

Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonExampleNew.scala]: set it with previous CarbonProperties in line 46

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonExampleNew.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      .addProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT, "yyyy/MM/dd")
+
+    val path = s"$rootPath/examples/spark/src/main/resources/data.csv"
+
+    spark.sql("DROP TABLE IF EXISTS t3")
+
+    // Create table, 6 dimensions, 1 measure
+    spark.sql("""
+           CREATE TABLE IF NOT EXISTS t3
+           (ID Int, date Timestamp, country String,
```
Comment (jackylk) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonExampleNew.scala]: can you use the same data set as in CarbonExample?

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/AbstractCarbonSparkSQLParser.scala [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  protected implicit def asParser(k: Keyword): Parser[String] = k.parser
+
+  // By default, use Reflection to find the reserved words defined in the sub class.
+  // NOTICE, Since the Keyword properties defined by sub class, we couldn't call this
+  // method during the parent class instantiation, because the sub class instance
+  // isn't created yet.
+  protected lazy val reservedWords: Seq[String] =
+  this
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/AbstractCarbonSparkSQLParser.scala]: incorrect indentation

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonCommonSqlParser.scala [hunk 5]
```diff
@@ -0,0 +1,970 @@
...
+   * This will extract the no inverted columns fields.
+   * By default all dimensions use inverted index.
+   *
+   * @param fields
+   * @param tableProperties
+   * @return
+   */
+  protected def extractNoInvertedIndexColumns(fields: Seq[Field],
+      tableProperties: Map[String, String]):
+  Seq[String] = {
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonCommonSqlParser.scala]: move to previous line

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonCommonSqlParser.scala [hunk 6]
```diff
@@ -0,0 +1,970 @@
...
+    // check whether the column name is in fields
+    var noInvertedIdxColsProps: Array[String] = Array[String]()
+    var noInvertedIdxCols: Seq[String] = Seq[String]()
+
+    if (tableProperties.get("NO_INVERTED_INDEX").isDefined) {
+      noInvertedIdxColsProps =
+        tableProperties.get("NO_INVERTED_INDEX").get.split(',').map(_.trim)
+      noInvertedIdxColsProps
+        .map { noInvertedIdxColProp =>
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonCommonSqlParser.scala]: move to previous line

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonCommonSqlParser.scala [hunk 7]
```diff
@@ -0,0 +1,970 @@
...
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.DataTypeUtil
+import org.apache.carbondata.spark.exception.MalformedCarbonCommandException
+import org.apache.carbondata.spark.util.CommonUtil
+
+/**
+ * TODO remove the duplicate code and add the common methods to common class.
+ * Parser for All Carbon DDL, DML cases in Unified context
+ */
+abstract class CarbonCommonSqlParser extends AbstractCarbonSparkSQLParser {
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonCommonSqlParser.scala]: Suggest to change to CarbonDDLSqlParser

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonCommonSqlParser.scala [hunk 8]
```diff
@@ -0,0 +1,970 @@
...
+import org.apache.carbondata.common.logging.LogServiceFactory
+import org.apache.carbondata.core.carbon.metadata.datatype.DataType
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+import org.apache.carbondata.core.util.DataTypeUtil
+import org.apache.carbondata.spark.exception.MalformedCarbonCommandException
+import org.apache.carbondata.spark.util.CommonUtil
+
+/**
+ * TODO remove the duplicate code and add the common methods to common class.
+ * Parser for All Carbon DDL, DML cases in Unified context
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonCommonSqlParser.scala]: DML is not in this parser scope, right?

**Comment by jackylk** [comment]
Can you also add ThriftServer support for CarbonSession?

**Comment by ravipesala** [comment]
@jackylk Fixed review comments, Please review.
I will raise another PR to add thrift server.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/329/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/330/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**